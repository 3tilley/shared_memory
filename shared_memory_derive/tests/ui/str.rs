//! Since only `str` implements SharedMemCast, it should be impossible to store a `&str` in
//! a struct and then derive SharedMemCast. This would be invalid anyway since you can't store
//! references in shared memory.

use shared_memory::SharedMemCast;

#[derive(SharedMemCast)]
pub struct MyStruct<'a> {
    x: &'a str,
}

#[derive(SharedMemCast)]
pub struct MyStruct2<'a>(&'a str);

// No lifetime on struct itself so forced to reject based on field
#[derive(SharedMemCast)]
pub struct MyStructStatic {
    x: &'static str,
}

#[derive(SharedMemCast)]
pub struct MyStruct2Static(&'static str);

fn main() {}
