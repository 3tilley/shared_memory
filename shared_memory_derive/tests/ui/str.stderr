error[E0277]: the trait bound `&'a str: shared_memory::cast::SharedMemCast` is not satisfied
 --> $DIR/str.rs:9:8
  |
9 |     x: &'a str,
  |        ^
  |        |
  |        the trait `shared_memory::cast::SharedMemCast` is not implemented for `&'a str`
  |        help: consider removing 1 leading `&`-references
  |
  = help: the following implementations were found:
            <str as shared_memory::cast::SharedMemCast>
  = note: required by `shared_memory::cast::AssertIsSharedMemCast`

error[E0277]: the trait bound `&'a str: shared_memory::cast::SharedMemCast` is not satisfied
  --> $DIR/str.rs:13:26
   |
13 | pub struct MyStruct2<'a>(&'a str);
   |                          ^
   |                          |
   |                          the trait `shared_memory::cast::SharedMemCast` is not implemented for `&'a str`
   |                          help: consider removing 1 leading `&`-references
   |
   = help: the following implementations were found:
             <str as shared_memory::cast::SharedMemCast>
   = note: required by `shared_memory::cast::AssertIsSharedMemCast`

error[E0277]: the trait bound `&'static str: shared_memory::cast::SharedMemCast` is not satisfied
  --> $DIR/str.rs:18:8
   |
18 |     x: &'static str,
   |        ^
   |        |
   |        the trait `shared_memory::cast::SharedMemCast` is not implemented for `&'static str`
   |        help: consider removing 1 leading `&`-references
   |
   = help: the following implementations were found:
             <str as shared_memory::cast::SharedMemCast>
   = note: required by `shared_memory::cast::AssertIsSharedMemCast`

error[E0277]: the trait bound `&'static str: shared_memory::cast::SharedMemCast` is not satisfied
  --> $DIR/str.rs:22:28
   |
22 | pub struct MyStruct2Static(&'static str);
   |                            ^
   |                            |
   |                            the trait `shared_memory::cast::SharedMemCast` is not implemented for `&'static str`
   |                            help: consider removing 1 leading `&`-references
   |
   = help: the following implementations were found:
             <str as shared_memory::cast::SharedMemCast>
   = note: required by `shared_memory::cast::AssertIsSharedMemCast`

For more information about this error, try `rustc --explain E0277`.
